rest.go

```diff
@@ -0,0 +1,333 @@
+		for _, key := range r.content[path].keys {
+			ctxKey := getContextKey(key)
+			str = fmt.Sprintf("r.Use(makeContextSaver(%s, \"%s\"))\n", ctxKey, key)
+			buffer.WriteString(str)
+		}
+		middleware := r.content[path].middleware
+		if middleware != "" {
+			buffer.WriteString(fmt.Sprintf("r.Use(m.%s()...)\n", middleware))
+		}
+		methods := r.content[path].methods
+		for _, m := range []string{"Get", "Post", "Put", "Delete"} {
+			handler, ok := methods[m]
+			if ok {
+				buffer.WriteString(fmt.Sprintf("r.%s(\"/\", rh.handle%s)\n", m, handler))
+			}
+		}
+		buffer.WriteString("})\n")
+	}
+	return buffer.String()
+}
+
+func getRoutes(app *pb.Application, epNames []string) (routes, error) {
+	paths := make([]string, 0, len(epNames)/2)
+	content := make(map[string]*route, len(epNames)/2)
+	for _, name := range epNames {
+		fields := strings.Split(name, " ")
+		if len(fields) != 2 {
+			msg := `expect "GET|POST|etc path/path" as endpoint name (%s) `
+			return routes{}, fmt.Errorf(msg, name)
+		}
+		httpMethod := strings.Title(strings.ToLower(fields[0]))
+		if _, ok := validHTTPMethods[httpMethod]; !ok {
+			return routes{}, fmt.Errorf("invalid HTTP Method (%s)", httpMethod)
+		}
+		endpoint := app.Endpoints[name]
+		httpPath := fields[1]
+		if _, ok := content[httpPath]; !ok {
+			middleware := ""
+			if m, ok := endpoint.Attrs["middleware"]; ok {
+				middleware = m.GetS()
+			}
+			content[httpPath] = &route{
+				methods:     map[string]string{},
+				middleware:  middleware,
+				keys:        getPatternParams(endpoint),
+				queryParams: getQueryParams(endpoint),
+			}
+			paths = append(paths, httpPath)
+		}
+		interfaceMethod := GenMethodName(endpoint)
+		content[httpPath].methods[httpMethod] = interfaceMethod
+		if httpMethod == "Put" {
+			t := getPayloadType(endpoint)
+			content[httpPath].putPayloadType = t
+		}
+		if httpMethod == "Post" {
+			content[httpPath].postPayloadType = getPayloadType(endpoint)
+		}
+	}
+	return routes{paths, content}, nil
+}
+
+func getPayloadType(ep *pb.Endpoint) string {
+	if len(ep.Param) != 1 || ep.Param[0].Type.GetTypeRef() == nil {
+		return ""
+	}
+	return ep.Param[0].Type.GetTypeRef().Ref.Appname.Part[0]
+}
+
+func genContextKeys(app *pb.Application, epNames []string) string {
+	keys := getContextKeys(app, epNames)
+	if len(keys) == 0 {
+		return ""
+	}
+	var buffer bytes.Buffer
+	buffer.WriteString("// Keys for Context lookup\nconst (\n")
+	s := fmt.Sprintf("%s ContextKeyType = iota\n", getContextKey(keys[0]))
+	buffer.WriteString(s)
+	for i := 1; i < len(keys); i++ {
+		buffer.WriteString(getContextKey(keys[i]) + "\n")
+	}
+	buffer.WriteString(")\n")
+	return buffer.String()
+}
+
+func getContextKey(param string) string {
+	return strings.Title(param) + "Key"
+}
+
+func getPatternParams(ep *pb.Endpoint) []string {
+	return getParams(ep, false)
+}
+
+func getQueryParams(ep *pb.Endpoint) []string {
+	return getParams(ep, true)
+}
+
+func getParams(ep *pb.Endpoint, queryType bool) []string {
```

Comment:
What does `queryType` mean? It's name should be in logical proposition form, e.g.: `wantQueryType` (assuming that's what was meant).

Reply:
Got rid of it altogether as discussed.